internal/ipfamily/ipfamily.go

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package ipfamily // import "go.universe.tf/metallb/internal/ipfamily"
+
+import (
+	"fmt"
+	"net"
+)
+
+// IP family helps identifying single stack IPv4/IPv6 vs Dual-stack ["IPv4", "IPv6"] or ["IPv6", "Ipv4"].
+type IPFamily string
+
+const (
+	IPv4      IPFamily = "ipv4"
+	IPv6      IPFamily = "ipv6"
+	DualStack IPFamily = "dual"
+	Invalid   IPFamily = "invalid"
+)
+
+// IPFamilyForAddresses returns the address family given list of addresses strings.
+func IPFamilyForAddresses(ips []string) (IPFamily, error) {
+	switch len(ips) {
+	case 1:
+		ip := net.ParseIP(ips[0])
+		if ip.To4() != nil {
+			return IPv4, nil
+		} else {
+			return IPv6, nil
+		}
+	case 2:
+		ip1 := net.ParseIP(ips[0])
+		ip2 := net.ParseIP(ips[1])
+		if ip1 == nil || ip2 == nil {
+			return Invalid, fmt.Errorf("IPFamilyForAddresses: Invalid address %q", ips)
+		}
+		if (ip1.To4() == nil) == (ip2.To4() == nil) {
+			return Invalid, fmt.Errorf("IPFamilyForAddresses: same address family %q", ips)
+		}
+		return DualStack, nil
+	default:
+		return Invalid, fmt.Errorf("IPFamilyForAddresses: invalid ips length %d %q", len(ips), ips)
+	}
+}
+
+// IPFamilyForAddressesIPs returns the address family from a given list of addresses IPs.
+func IPFamilyForAddressesIPs(ips []net.IP) (IPFamily, error) {
```

Comment:
I'd convert ips to []string and use the one above, so the logic is in one place only.